# 06-02 · Poruszanie się do przodu i do tyłu

Praktyka do sekcji [„Moving forward and backward”
(Rozdział 6 · Turtle książka „Python od zera”

## Cel

Nauczyć się sterować żółwiem Turtle: `forward`, `backward`, `right`, `left` —
i w praktyce odczuwać różnicę między **ruchem** a **obrotem**.

## Co trzeba wiedzieć

Ten laptop to laboratorium do części teorii, którą już czytałem. Jeśli tu przyszedłeś
po raz pierwszy, krótko: żółw ma pozycję i kurs (kierunek). `forward`/`backward`
przesuwać go w linii prostej, `left`/`right` tylko skręcać trasę.

## Krótkie przypomnienie

- `artist.forward(n)` / `artist.fd(n)` — naprzód `n` pikseli
- `artist.backward(n)` / `artist.bk(n)` / `artist.back(n)` — powrót do `n` pikseli
- `artist.left(угол)` / `artist.lt(угол)` – obróć się przeciwnie do ruchu wskazówek zegara
- `artist.right(угол)` / `artist.rt(угол)` – obróć zgodnie z ruchem wskazówek zegara

**Cechy pracy w Jupyter:** okno Turtle otwiera się raz na cały notebook —
tworzymy `screen` oraz `artist` w następnej celi. Przed każdym nowym rysunkiem
wywołujemy `artist.reset()`oczyścić płótno i przywrócić żółwia na początek
współrzędne; zamknąć okno za pomocą komendy `screen.bye()` tylko w ostatniej celi.
(Technicznie rzecz biorąc, nie jest możliwe utworzenie i zamknięcie osobnego okna Turtle w każdej komórce —
moduł `turtle` nie jest zaprojektowany do tego w ramach jednego procesu.)

## Ustawienie (wykonać raz)

In [1]:
import turtle

screen = turtle.Screen()
artist = turtle.Turtle()
print("Окно Turtle готово. Позиция:", artist.position(), "| курс:", artist.heading())

Окно Turtle готово. Позиция: (0.00,0.00) | курс: 0.0


## Sprawa robocza

In [2]:
artist.reset()

# przesunąć do przodu o 120 pikseli
artist.forward(120)

# obracamy o 90 stopni przeciwnie do ruchu wskazówek zegara
artist.left(90)

# znowu jedziesz do przodu – dostajesz literę „G”
artist.forward(80)

print("Позиция:", artist.position())
print("Курс:", artist.heading())

Позиция: (120.00,80.00)
Курс: 90.0


## Eksperyment 1

Zmień odległość `120` dalej `250` i kąt `90` dalej `45`. Zanim zaczniesz, spróbuj
przewidzieć: w którą stronę żółw „patrzy”

In [3]:
artist.reset()

artist.forward(250)   # было 120
artist.left(45)       # было 90
artist.forward(80)

print("Курс после поворота:", artist.heading(), "градусов")

Курс после поворота: 45.0 градусов


## Eksperyment 2

Dodaj trzecią parę poleceń `left(90)` oraz `forward(120)` na końcu programu. Jakie
jaka figura powstanie?

In [4]:
artist.reset()

artist.forward(120)
artist.left(90)
artist.forward(80)
artist.left(90)        # новый поворот
artist.forward(120)    # новая сторона

print("Позиция:", artist.position())
print("Курс:", artist.heading())

Позиция: (0.00,80.00)
Курс: 180.0


## Podstawowa praktyka zadań ★

Używanie tylko `forward`, `backward`, `left` oraz `right`, narysuj literę „T” jednym
ciągłą linię pióra: najpierw górny pręt, potem noga w dół od środka.

Jedno z rozwiązań znajduje się poniżej. Spróbuj napisać własną wersję tego, jak ****
uruchomić tę komórkę.

In [5]:
artist.reset()

# górna poprzeczka: najpierw prawa połowa, potem przez środek – lewa
artist.forward(40)
artist.backward(80)
artist.forward(40)  # возвращаемся точно в центр перекладины

# trzonka litery „T”
artist.right(90)
artist.forward(100)

print("Позиция:", artist.position())
print("Курс:", artist.heading())

Позиция: (0.00,-100.00)
Курс: 270.0


## Sprawdzenie wyniku

Po wykonaniu celu zadania żółw musi zakończyć ruch w tym punkcie
`(0.0, -100.0)` z kursem `270°` (patrzy w dół) to dolna część łodygi litery „T”

Następna komórka powtarza rozwiązanie i automatycznie sprawdza ostateczny stan.

In [6]:
artist.reset()

artist.forward(40)
artist.backward(80)
artist.forward(40)
artist.right(90)
artist.forward(100)

x, y = artist.position()
assert round(x, 1) == 0.0, f"Неожиданная координата x: {x}"
assert round(y, 1) == -100.0, f"Неожиданная координата y: {y}"
assert artist.heading() == 270.0, f"Неожиданный курс: {artist.heading()}"

print("Проверка пройдена: буква «Т» нарисована корректно.")

Проверка пройдена: буква «Т» нарисована корректно.


## Typowy błąd

Początkujący często się mylą `right()`/`left()` z ruchem – wydaje się, że żółw
powinien przesunąć się w bok. W rzeczywistości te polecenia **tylko obracają** kurs
i nie poruszają żółwia ani o piksel. Poniżej — program z tym błędem: zamiast litery
„G” powstaje po prostu jako prosta linia o długości 200, ponieważ obrót „zginął” pomiędzy
dwa identyczne ruchy do przodu.

In [7]:
artist.reset()

artist.forward(120)
# błąd: zapomniano obrócić tutaj
artist.forward(80)

print("Курс (не изменился!):", artist.heading())
print("Черепашка проехала прямо, а не буквой «Г»:", artist.position())

Курс (не изменился!): 0.0
Черепашка проехала прямо, а не буквой «Г»: (200.00,0.00)


## Poprawka

Dodaj nieudane zwroty między dwoma częściami:

In [8]:
artist.reset()

artist.forward(120)
artist.left(90)   # добавили поворот
artist.forward(80)

print("Курс после исправления:", artist.heading())
print("Позиция:", artist.position())

Курс после исправления: 90.0
Позиция: (120.00,80.00)


## Praktyka samodzielne

Narysuj schody z trzech identycznych „stopni”
powtórzyć trzykrotnie). Dopóki nie poznano pętli (rozdział 10), po prostu powtórz potrzebny blok
trzy polecenia z rzędu wykonuje ręcznie.

In [9]:
artist.reset()

# jeden krok – dodaj jeszcze dwa według tego samego przykładu
artist.forward(40)
artist.left(90)
artist.forward(40)
artist.right(90)

print("Позиция после одной ступеньки:", artist.position())

Позиция после одной ступеньки: (40.00,40.00)


## Zadanie dodatkowe ★★★

Narysuj pięcioramienną gwiazdę jedną linią, używając tylko `forward` i jeden i drugi
ten sam kąt obrotu między wszystkimi pięcioma ruchami. Podpowiedź: Suma zewnętrznych narożników
suma kątów wielokąta zawsze wynosi 360° — dla pięcioramiennej gwiazdy ten kąt wynosi 144°.

In [10]:
artist.reset()

for _ in range(5):
    artist.forward(150)
    artist.right(144)

print("Звезда нарисована. Финальная позиция:", artist.position())

Звезда нарисована. Финальная позиция: (-0.00,-0.00)


## Ukończenie (wykonaj raz, na samym końcu)

In [11]:
screen.bye()
print("Окно Turtle закрыто.")

Окно Turtle закрыто.
